### PR TITLE
feat(core): app menu waffle launcher

### LINKED ISSUE #59948: App navigation: waffle menu and header layout
Sub-issue of #59888.

Replace the current `AppMenu` (horizontal flex with overflow dropdown) with a popover-style waffle launcher. Apps live in a 4-column grid with theme-colored circle backgrounds, capped at 3 visible rows with scrolling beyond.

Two triggers open the popover: the waffle icon and the current-app icon+name. 

Smaller things included:
- Admin-only "More apps" entry pointing at `/settings/apps`
- Apps open in a new tab
- Reusable app link/icon component (Dashboard will reuse it later)
- Keyboard nav across the grid, focus trap, focus return on close

### PR DESCRIPTION
**Pull request by pringelmann**
* Resolves: #59948

## Summary

First pass at the waffle launcher behind the apps icon in the header. Opening it shows a 4-wide grid of installed apps in a popover; admins also see a "More apps" tile linking to the app store. The existing current-app button is kept as a second trigger for the same popover.

[Preview link](https://nc-preview.tail22c02b.ts.net/)

image

What's in:

- Popover grid replaces the old hamburger overlay
- Roving tabindex keyboard nav (arrow keys, Home/End, Enter/Space)
- Focus returns to whichever trigger opened the popover
- RTL-aware horizontal alignment
- Admin-only "More apps" tile
- Unit + Cypress coverage

## TODO

Follow-ups for later PRs:

- [ ] Live unread/notification counts on the tiles
- [ ] Drag-and-drop reordering, persisted per user
- [ ] Broaden test coverage (RTL, focus trap edge cases, keyboard variants, viewport breakpoints)

## Checklist

- [x] Code is properly formatted
- [x] Sign-off message is added to all commits
- [x] Tests (unit, integration, api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes (preview link)
- [ ] Documentation (manuals or wiki) has been updated or is not required
- [ ] Backports requested where applicable (ex: critical bugfixes)
- [x] Labels added where applicable (ex: bug/enhancement, `3. to review`, feature component)
- [x] Milestone added for target branch/version (ex: 32.x for `stable32`)

### CODE REVIEW

File: core/src/components/AppItem.vue [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+
+<script setup lang="ts">
+import type { INavigationEntry } from '../types/navigation.d.ts'
+
+import { n } from '@nextcloud/l10n'
+import { computed } from 'vue'
+
+const props = withDefaults(defineProps<{
+	app: INavigationEntry
+	/** When true, the link opens in a new tab. Launcher uses true; Dashboard reuse will use false. */
```
Comment (pringelmann) [inline core/src/components/AppItem.vue]: This behaviour is still TBC. We need to settle on a final UX solution before releasing this.

File: core/src/components/AppItem.vue [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+		'{count} notification',
+		'{count} notifications',
+		props.app.unread,
+		{ count: props.app.unread },
+	)
+})
+</script>
+
+<style scoped lang="scss">
+.app-item {
```
Comment (pringelmann) [inline core/src/components/AppItem.vue:75]: We can consider moving this to `@nextcloud/vue` in future, since this item will be used on dashboard, universal search, etc. 

Would like to ship in-place for first iteration.

File: core/src/components/AppMenu.vue [hunk 3]
```diff
@@ -4,161 +4,447 @@
...
 		const appList = loadState<INavigationEntry[]>('core', 'apps', [])
 		return {
 			appList,
+			isAdmin: getCurrentUser()?.isAdmin ?? false,
+			// Roving tabindex: only this tile has tabindex=0; arrow keys move it.
+			focusedIndex: 0,
+			// NcPopover's focus-trap only knows the slot trigger (waffle).
+			// The current-app button lives outside the slot, so we track the
+			// source and restore focus manually via setReturnFocus.
+			openedFrom: null as 'waffle' | 'currentApp' | null,
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:114]: UX behaviour relevant question here. Used common sense for first iteration.

File: core/src/components/AppMenu.vue [hunk 4]
```diff
@@ -4,161 +4,447 @@
...
+			// via the browser's native modifier-aware <a> activation.
+			if (event.ctrlKey || event.metaKey || event.altKey || event.shiftKey) {
+				return
+			}
+
+			if (this.gridItems.length === 0) {
+				return
+			}
+
+			const cols = 4
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:243]: magic number 😬 imo fine for first iteration, but some rework might be needed here.

File: core/src/components/AppMenu.vue [hunk 5]
```diff
@@ -4,161 +4,447 @@
...
+					next = 0
+					break
+				case 'End':
+					next = total - 1
+					break
+				case 'Enter':
+				case ' ': {
+					// Space's default scrolls the nearest scrollable ancestor (the
+					// popover); intercept and click programmatically. Enter gets
+					// the same treatment so we can close the popover uniformly.
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:286]: Consistent with other waffle menus I looked at, but worth flagging as UX question

File: core/src/components/AppMenu.vue [hunk 6]
```diff
@@ -4,161 +4,447 @@
...
-	width: 0;
+	align-items: center;
 
-	&__list {
+	&__waffle {
+		// NcButton's tertiary-no-background variant uses --color-main-text,
+		// which is dark on light themes. The header sits on the theme primary
+		// background, so override to use the matching plain-text color.
+		--color-main-text: var(--color-background-plain-text);
+		color: var(--color-background-plain-text);
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:325]: Might be better to add variants to `@nextcloud/vue`, but imo fine for first iteration

File: core/src/components/AppMenu.vue [hunk 7]
```diff
@@ -4,161 +4,447 @@
...
-					opacity: 1;
-					background-color: transparent !important;
-				}
-			}
+	&__grid {
+		--app-item-col-width: 69px;
+		--app-item-row-height: 64px;
+		--app-menu-rows-visible: 6;
+		padding: calc(var(--default-grid-baseline) * 3) calc(var(--default-grid-baseline) * 2);
+		display: grid;
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:396]: We can debate whether grid is best/simplest solution here. Happy to iterate on this.

File: core/src/components/AppMenu.vue [hunk 8]
```diff
@@ -4,161 +4,447 @@
...
+}
+
+// Outside the scoped block: ::-webkit-scrollbar pseudo-elements need unscoped
+// CSS to bind in Chrome. !important: core/css/styles.scss forces a 12 px thumb.
+.app-menu__popover-base .app-menu__grid {
+	scrollbar-width: thin !important;
+	scrollbar-color: var(--color-scrollbar) transparent !important;
+
+	&::-webkit-scrollbar {
+		width: 6px !important;
```
Comment (pringelmann) [inline core/src/components/AppMenu.vue:439]: Quick and dirty scrollbar styling. Happy to iterate on this.

File: core/src/components/AppItem.vue (hunk 2 above)
Comment (susnux) [reply core/src/components/AppItem.vue:75]: if planned for nextcloud-vue it would be great to use css modules instead of scoped styles for two-way isolation of styles.

File: core/src/components/AppItem.vue [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+	outlined?: boolean
+	/**
+	 * Roving-tabindex value. AppMenu sets this to 0 on the focused tile and
+	 * -1 on all other tiles so only one stop is in the natural Tab order.
+	 * Default -1 keeps tiles out of the Tab order when used standalone.
+	 */
+	tabindex?: number
+}>(), {
+	newTab: false,
+	outlined: false,
```
Comment (susnux) [inline core/src/components/AppItem.vue]: thats default

```suggestion
```

File: core/src/components/AppMenu.vue [hunk 10]
```diff
@@ -4,161 +4,459 @@
...
+						:newTab="item.id === 'app-store'"
+						:tabindex="i === focusedIndex ? 0 : -1" />
+				</div>
+			</div>
+		</NcPopover>
+		<button
+			v-if="currentApp"
+			class="app-menu__current-app"
+			type="button"
+			:aria-label="t('core', 'Open apps menu')"
```
Comment (susnux) [inline core/src/components/AppMenu.vue:52]: so both buttons open the waffle menu?
I would expect this button to open the index of the current app instead

File: core/src/components/AppMenu.vue (hunk 5 above)
Comment (susnux) [reply core/src/components/AppMenu.vue:286]: space seems a bit unexpected here rest seems fine (for accessibility)

File: core/src/components/AppMenu.vue (hunk 10 above)
Comment (pringelmann) [reply core/src/components/AppMenu.vue:52]: Yes as per [design spec]( https://github.com/nextcloud/server/issues/59888#:~:text=The%20icon%20for%20the%20app%20menu%20would%20be%20placed%20after%20the%20logo%2E%20The%20icon%20and%20name%20of%20the%20current%20app%20would%20be%20placed%20after%2E%20Clicking%20these%20would%20also%20pop%20up%20the%20app%20menu
)

This also feels like a weird interaction to me, especially since the popup anchor still points to the waffle. Would you mind expanding on what the rationale is here @kra-mo?

Comment (susnux) [reply core/src/components/AppMenu.vue:52]: agree UX wise this is unexpected

File: core/src/components/AppMenu.vue (hunk 5 above)
Comment (pringelmann) [reply core/src/components/AppMenu.vue:286]: fwiw APG recommends this behaviour:

https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/#:~:text=activates%20the%20menu%20item%2C

My reasoning is that `arrows`, `Home` and `End` all trigger focus events (which will auto-scroll), making it possible for users to scroll the list without needing a dedicated key for it.

File: core/src/components/AppMenu.vue (hunk 10 above)
Comment (kra-mo) [reply core/src/components/AppMenu.vue:52]: > This also feels like a weird interaction to me, especially since the popup anchor still points to the waffle. Would you mind expanding on what the rationale is here @kra-mo?

Since most apps that need it already provide a navigation item to go "Home", it was to make the menu more easy to access and discover.

I thought of both behaviors and this just felt "safer", even if more unconventional. Maybe I'm just being overly cautious though.

Were we to keep this, it would be good to link the two as a single button (when focused or hovered for example).

Comment (pringelmann) [reply core/src/components/AppMenu.vue:52]: Ok yeah that makes sense. I think a single button would definitely help yes. Perhaps we go with what we have for the first step and gather some internal feedback?

**Review by artonge — APPROVED** [review]

**Comment by artonge** [comment]
Approved to be merged for the screenshots, we will address all the raised concerns in a follow-up.

**Comment by schiessle** [comment]
Because truncated app names are a concern raised from users every now and than with the "old menu". Does this menu adjust the column width automatically so that the app name is always completely readable (also think about translation where app names might be longer)?

**Review by AndyScherzinger — APPROVED** [review]

**Comment by kra-mo** [comment]
> Does this menu adjust the column width automatically so that the app name is always completely readable

The menu should generally leave more room for app names by default and it should wrap them instead of ellipsizing:

image

**Comment by pringelmann** [comment]
> > Does this menu adjust the column width automatically so that the app name is always completely readable
> 
> The menu should generally leave more room for app names by default and it should wrap them instead of ellipsizing:
> image

This is how it behaves:

Screenshot_2026-05-07_13-08-22

**Comment by kra-mo** [comment]
Ha, sorry, by should, I meant it does. It was a screenshot :)

**Review by kra-mo — DISMISSED** [review]
Looks generally good as a first iteration from the design side. But apps should open on the same page as previously mentioned (I'm not sure if that has been changed already, I can't test it besides the preview page).

Issues I saw on the demo:

The + icon seems to be missing?

Image

When opening the menu from the app name button, it is left in this state after closing it:

Image

And a small nit:

Image

It would be good if the padding of the container was consistent all-around and instead, the top row had increased top padding. This way, the hover/focus effect would be concentric(-ish) with the outside container :)

**Review by kra-mo — COMMENTED** [review]
Ah, and my previous comment about the colors in dark mode: https://github.com/nextcloud/server/issues/59888#issuecomment-4335284514

**Comment by pringelmann** [comment]
> Looks generally good as a first iteration from the design side. But apps should open on the same page as previously mentioned (I'm not sure if that has been changed already, I can't test it besides the preview page).
> 
> Issues I saw on the demo:
> 
> The + icon seems to be missing?
> Image
> 
> When opening the menu from the app name button, it is left in this state after closing it:
> Image
> 
> And a small nit:
> Image
> 
> It would be good if the padding of the container was consistent all-around and instead, the top row had increased top padding. This way, the hover/focus effect would be concentric(-ish) with the outside container :)

Fixed already: (you might need to do a hard refresh)'

- Missing + icon
- Lingering white button focus state
- Links open in same tab 

Will look into padding issue

**Review by jancborchardt — APPROVED** [review]

**Comment by artonge** [comment]
CI failure unrelated
